gql/client.py

```diff
@@ -1038,6 +1184,105 @@ def execute_batch(
 
         return cast(List[Dict[str, Any]], [result.data for result in results])
 
+    def _batch_loop(self) -> None:
+        """main loop of the thread used to wait for requests
+        to execute them in a batch"""
+
+        stop_loop = False
+
+        while not stop_loop:
+
+            # First wait for a first request in from the batch queue
+            requests_and_futures: List[Tuple[GraphQLRequest, Future]] = []
+            request_and_future: Tuple[GraphQLRequest, Future] = self.batch_queue.get()
+            if request_and_future is None:
+                break
+            requests_and_futures.append(request_and_future)
+
+            # Then wait the requested batch interval except if we already
+            # have the maximum number of requests in the queue
+            if self.batch_queue.qsize() < self.client.batch_max - 1:
+                time.sleep(self.client.batch_interval)
+
+            # Then get the requests which had been made during that wait interval
+            for _ in range(self.client.batch_max - 1):
+                if self.batch_queue.empty():
+                    break
+                request_and_future = self.batch_queue.get()
+                if request_and_future is None:
+                    stop_loop = True
```

Comment:
Another possible option here is to wrap the while loop, with try - except, and raise an exception in here. If requests_and_futures is not empty, set_exception for each future. I'll leave the decision to you if you want to do it or not. I consider this way is also fine, if the user wants to perform the last requests after the termination.

Reply:
It is necessary to wait for all the requests to be made or else the promise could never resolve.